stream_alert/shared/lookup_tables/driver_s3.py

```diff
@@ -0,0 +1,195 @@
+        self._reload()
+
+    def commit(self):
+        if not self._dirty:
+            LOGGER.warn('LookupTable (%s): Empty commit; no records dirtied', self.id)
+            return
+
+        raise NotImplementedError('Help derek you screwed me')
+
+    def get(self, key, default=None):
+        self._reload_if_necessary()
+        return self._cache.get(key, default)
+
+    def set(self, key, value):
+        """
+        Under construction
+        FIXME (derek.wang)
+        """
+        self._dirty = True
+
+    def _reload_if_necessary(self):
+        """
+        Uses the "cache_refresh_minutes" option to determine whether or not the current LookupTable
+        should be re-fetched from S3.
+
+        If it needs a reload, this method will appropriately call reload.
+        """
+        if self._cache.has(self._MAGIC_CACHE_TTL_KEY):
+            LOGGER.debug(
+                'LookupTable (%s): Does not need refresh. TTL: %s',
+                self.id,
+                self._cache.ttl(self._MAGIC_CACHE_TTL_KEY)
+            )
+
+        else:
+            LOGGER.info(
+                'LookupTable (%s): Needs refresh, starting now.',
+                self.id
+            )
+            self._reload()
+
+    def _reload(self):
+        """
+        Reaches to AWS S3, downloads the relevant data file, decompresses, decodes, and caches
+        the file's contents into memory.
+        """
+        # First, download the item from S3
+        try:
+            start_time = time.time()
+            s3_object = self._s3_client.Object(self._s3_bucket, self._s3_key).get()
+            data = s3_object.get('Body').read()
+
+            total_time = time.time() - start_time
+            size_kb = round(s3_object.get('ContentLength') / 1024.0, 2)
+            size_mb = round(size_kb / 1024.0, 2)
+            LOGGER.debug(
+                'LookupTable (%s): Downloaded S3 file size %s in %s seconds',
+                self.id,
+                '{}MB'.format(size_mb) if size_mb else '{}KB'.format(size_kb),
+                round(total_time, 2)
+            )
+        except ClientError as err:
+            LOGGER.error(
+                'LookupTable (%s): Encountered error while downloading %s from %s: %s',
+                self.id,
+                self._s3_key,
+                self._s3_bucket,
+                err.response['Error']['Message']
+            )
+            raise LookupTablesInitializationError(
+                'LookupTable S3 Driver Failed with Message: {}'.format(
+                    err.response['Error']['Message']
+                )
+            )
+
+        except (ConnectTimeoutError, ReadTimeoutError):
+            # Catching ConnectTimeoutError and ReadTimeoutError from botocore
+            LOGGER.error(
+                'LookupTable (%s): Reading from S3 timed out',
+                self.id
+            )
+            raise LookupTablesInitializationError(
```

Comment:
Should we fail the whole lambda because of LookupTables errors especially timeout. It is common to see timeout? [In current LookupTables code](https://github.com/airbnb/streamalert/blob/master/stream_alert/shared/lookup_tables.py#L71-L78), we log the errors without failing the whole lambda.